### ISSUE TITLE
Media remove requests too many tokens with Swift backend when using concurrency

### ISSUE THREAD

**Issue by mrdaemon** [mastodon/mastodon#21602]
### Steps to reproduce the problem

1. Use Swift Object storage (I use OVH with v3 auth, as it is the only configuration that works in my geo zone)
2. Accumulate many thousands of media files from other instances over time
3. run `RAILS_ENV=production <...>/bin/tootctl media remove` to remove the ones older than 7 days
4. Receive 429 Too Many Requests very quickly and operation completes without actually deleting most of the files.

### Expected behaviour

Old cached media files on Swift Object Storage backend should have been cleaned up

### Actual behaviour

429 Too many Requests

### Detailed description

Objects start deleting but soon:

```
Error processing 109357407426597968: Expected([200, 201]) <=> Actual(429 Too Many Requests)
excon.error.response
  :body          => ""
  :cookies       => [
  ]
  :headers       => {
    "Connection"        => "close"
    "X-IPLB-Instance"   => "<snip snip>"
    "X-IPLB-Request-ID" => "<snip snip>"
  }
  :host          => "auth.cloud.ovh.net"
  :local_address => "<snip snip>"
  :local_port    => 40360
  :path          => "/v3/auth/tokens"
  :port          => 443
  :reason_phrase => "Too Many Requests"
  :remote_ip     => "<snip snip>"
  :status        => 429
  :status_line   => "HTTP/1.1 429 Too Many Requests\r\n"
```

The auth token endpoint starts rejecting the requests as too many tokens have been requested in a short time span.

Lowering the concurrency to 2 delays the issue but it eventually returns, disabling it (`--concurrency=1`) is a functional workaround but it is taking about 30 earth ages with an ETA of the heat death of the universe, give or take, especially since the job has never ran since the instance's inception, and the cached media files have been accumulating.

The [OVH documentation](https://docs.ovh.com/ca/en/storage/object-storage/pcs/capabilities-and-limitations/) states that a token is valid for 24 hours, at first glance.

I am not too well versed in the implementation details of Swift, to be quite honest, but couldn't this be batched without requesting new tokens for every thread/worker? Again, I have no idea how this is implemented so feel free to disregard. If it is a limitation that is absolutely fine, and the job __will__ run with No Concurrency, but that's gonna start being awful when the instance grows, as the process is obnoxiously slow.

### Specifications

Mastodon 4.0.2
Ruby 3.0.4p208
Node.js v16.18.1

OVH Public Cloud Swift Object Storage via v3 auth api

Not running in docker container.

**Comment by brookmiles** [comment]
This is becoming a problem for me as well, since without concurrency daily cleanup is beginning to take as long as 10-12 hours, and I would expect this may continue to increase.

**Comment by brookmiles** [comment]
I've worked around this problem for now by manually patching `./vendor/bundle/ruby/3.0.0/gems/kt-paperclip-7.1.1/lib/paperclip/storage/fog.rb` to add credential caching, based on https://github.com/thoughtbot/paperclip/pull/2386 and whatever I've learned about ruby in the past hour.

```ruby
      def connection
#        @connection ||= ::Fog::Storage.new(fog_credentials)
        instances = (Thread.current[:paperclip_fog_instances] ||= ActiveSupport::Cache::MemoryStore.new(expires_in: 5.minutes))
        instances.fetch(fog_credentials) do
          warn("creating new fog connection for #{Thread.current}")
          ::Fog::Storage.new(fog_credentials)
        end
      end
```

**Comment by mrdaemon** [comment]
Quick update to mention that as of Mastodon 4.1.0, the above patch to paperclip has unintended side effects and causes sidekiq workers to raise an exception. I suspect some ActiveSupport cache semantic change , but i don't have the Ruby knowledge to fix it myself.

```
 TypeError: no _dump_data is defined for class Thread
  WARN: /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.7.2/lib/active_support/cache.rb:841:in `dump'
  /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.7.2/lib/active_support/cache.rb:841:in `dup_value!'
  /home/mastodon/live/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.7.2/lib/active_support/cache/memory_store.rb:36:in `dump'
  ```
 I've momentarily reverted it, will update if I come up with anything in case anyone ventures in this Issue looking for riches

**Comment by mrdaemon** [comment]
As an update, I've elected to try storing the connection object directly in the thread to be shared for subsequent calls within a single thread, without involving rails/ActiveSupport cache with expiration, as this now generates _something_ within the Thread object that is no longer serializable, and I have no idea what. Digging into fog-openstack only served to make me furrow my brow.

This mirrors the implementation in the upstream PR on [thoughtbot/paperclip](https://github.com/thoughtbot/paperclip/pull/2386).

```diff
--- vendor/bundle/ruby/3.0.0/gems/kt-paperclip-7.1.1/lib/paperclip/storage/fog.rb       2022-11-11 20:49:39.352053018 -0500
+++ vendor/bundle/ruby/3.0.0/gems/kt-paperclip-7.1.1/lib/paperclip/storage/fog.rb       2022-12-01 09:29:21.891686933 -0500
@@ -229,7 +229,8 @@
       end

       def connection
-        @connection ||= ::Fog::Storage.new(fog_credentials)
+        instances = (Thread.current[:paperclip_fog_instances] ||= {})
+        instances[fog_credentials] ||= ::Fog::Storage.new(fog_credentials)
       end

       def directory
```

I'm not 100% on it and I would deeply appreciate insight from anyone who is more familiar with the inner workings but from what I have been able to glean through the code and observing behavior, my rationale is:

* Paperclip/Fog calls seem to be only done through sidekiq tasks
* Sidekiq tasks use threads (one per) but they are not recycled, new ones seem spawned for every task (according to TID values in sidekiq and visible behavior)
* The auth token is valid for 24h
* None of the code instantiating the Fog connection has a lifetime that is > 24h at first glance.

This is not ideal, but after testing for 40 hours on my own instance while monitoring logs closely, it absolutely seems to work without ill effects. None of the connections failed, object storage works, and the initial problem has been circumvented as `tootctl media clean` seems to have long lived threads and the connection is then reused instead of re-established for every file.

IN SHORT if anyone ventures in this thread looking for a workaround on Swift object storage, this seems to function adequately so far.

**Comment by mrdaemon** [comment]
Last update on this, as I've also finally migrated storage to S3 compatible -- the above patch still works as of Mastodon 4.2.1, if anyone ventures in here looking for riches

Swift support is not ideal and I would recommend anyone applying this in production does the same and just use an S3 compatible object storage solution, since swift is deprecated and poorly supported by everything anyways

**Comment by mjankowski** [comment]
The original issue here is on an older mastodon version, and the last comment to me sort of implies this is a wontfix ... going to close on assumption there's not really much we can do here (all the "S3 compatible" things are somewhat out of control) ... please reopen if I'm missing some obvious remaining todo here!
